### PR TITLE
QRDQN implementation

### PR DESCRIPTION
**Pull request by Mobius1D**
QRDQN implementation with a few bugs that needs to be resolved.

Implementation details:
image

credits: https://arxiv.org/pdf/1710.10044.pdf

### CODE REVIEW

File: src/algorithms/dqns/qr_dqn.jl [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    batch_size = learner.sampler.batch_size
+    N = learner.ensemble_num
+    κ = learner.κ
+    τ = learner.τ
+    D = device(Q)
+
+    s, a, r, t, s′ = (send_to_device(D, batch[x]) for x in SARTS)
+    a = CartesianIndex.(repeat(batch.action, inner = N), 1:(N*batch_size))
+
+    target_q = reshape(Qₜ(s′), :, N, batch_size)
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: So here the return of `Qₜ(s′)` is expected to be of size `(n_action, n_quantile, batch_size)` ?

**Review by findmyway — CHANGES_REQUESTED** [review]

File: src/algorithms/dqns/qr_dqn.jl [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        q = reshape(q, 1, N, batch_size)
+
+        TD_error = (target .- q)
+        temp = Zygote.dropgrad(abs.(TD_error) .<  κ)
+        element_wise_huber_loss = ((TD_error.^2) .* temp) + κ*(TD_error .- 0.5*κ) .* (1 .- temp)
+
+        # dropgrad
+        element_wise_huber_loss =
+            abs.(reshape(τ, 1, N) .- Zygote.dropgrad(TD_error .< 0)) .*
+            element_wise_huber_loss ./ κ
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: better to move the quntile huberloss part into a funtion

File: src/algorithms/dqns/qr_dqn.jl (hunk 1 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Yes, originally it is of size (n_action*n_quantile, batch_size) and it is reshaped to (n_action, n_quantile, batch_size) for convenience.

File: src/algorithms/dqns/qr_dqn.jl (hunk 2 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: But it is the same when implemented without a function too right? But yea, can do that if it helps re-usability.

File: src/algorithms/dqns/qr_dqn.jl (hunk 1 above)
Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: Since Julia is column-major, better to design it as `(n_quantile, n_action, batch_size)` for performance and easier review.

File: src/algorithms/dqns/qr_dqn.jl (hunk 2 above)
Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: At least for your current implementation, there's a slight difference in `Zygote.gradient` .

File: src/algorithms/dqns/qr_dqn.jl [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+} <: AbstractLearner
+    approximator::Tq
+    target_approximator::Tt
+    loss_func::Tf
+    min_replay_history::Int
+    update_freq::Int
+    update_step::Int
+    target_update_freq::Int
+    sampler::NStepBatchSampler
+    ensemble_num::Int
```
Comment (pilgrimygy) [inline src/algorithms/dqns/qr_dqn.jl]: Maybe you could change `ensemble_num` to `quant_num`.

Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Will do that.

**Comment by Mobius1D** [comment]
I think there is something wrong with the loss function that I wasn't able to figure out. It would be great if anyone can help with that.

File: src/algorithms/dqns/qr_dqn.jl (hunk 1 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Makes sense, will do that.

File: src/algorithms/dqns/qr_dqn.jl (hunk 2 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Done.

File: src/algorithms/dqns/qr_dqn.jl (hunk 1 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Yes, there is a significant reduction in training time.

File: src/algorithms/dqns/qr_dqn.jl [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+y -> begin
+    x = @set x.approximator = y.Q
+    x = @set x.target_approximator = y.Qₜ
+    x
+end
+
+function (learner::QRDQNLearner)(env)
+    s = send_to_device(device(learner.approximator), state(env))
+    s = Flux.unsqueeze(s, ndims(s) + 1)
+    q = reshape(learner.approximator(s), :, learner.quantile_num)
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: The order is incorrect. Always remembet that in Julia the matrix is column major.

File: src/algorithms/dqns/qr_dqn.jl [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    vec(mean(q, dims = 2)) |> send_to_host
+end
+
+function quantile_huber_loss(target::AbstractArray, q::AbstractArray, κ::Float32, τ::Vector{Float64}, N::Int)
+    TD_error = (target .- q)
+
+    # dropgrad
+    temp = Zygote.dropgrad(abs.(TD_error) .<  κ)
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: There shouldn't be anything related to Zygote in loss function.

File: src/algorithms/dqns/qr_dqn.jl [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+end
+
+function (learner::QRDQNLearner)(env)
+    s = send_to_device(device(learner.approximator), state(env))
+    s = Flux.unsqueeze(s, ndims(s) + 1)
+    q = reshape(learner.approximator(s), :, learner.quantile_num)
+    vec(mean(q, dims = 2)) |> send_to_host
+end
+
+function quantile_huber_loss(target::AbstractArray, q::AbstractArray, κ::Float32, τ::Vector{Float64}, N::Int)
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: It won't work on GPU if you specfy \tau to be Vector

Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: Coulf you add a test case forthis function?

File: src/algorithms/dqns/qr_dqn.jl [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    avg_q = mean(target_q, dims=1)
+
+    if haskey(batch, :next_legal_actions_mask)
+        l′ = send_to_device(D, batch[:next_legal_actions_mask])
+        avg_q .+= ifelse.(l′, 0.0f0, typemin(Float32))
+    end
+
+    aₜ = argmax(avg_q, dims=2)
+    aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:N-1, 0:0, 0:0)))
+    qₜ = reshape(target_q[aₜ], :, batch_size)
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: use view here

**Review by findmyway — CHANGES_REQUESTED** [review]
Could you als paste the result of te experiment you added here?

File: src/algorithms/dqns/qr_dqn.jl (hunk 7 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Can you be a bit more specific? is it in the reshape or in aₜ?

File: src/algorithms/dqns/qr_dqn.jl (hunk 6 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Sure.

File: src/algorithms/dqns/qr_dqn.jl (hunk 7 above)
Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: I mean `target_q[aₜ]`

File: src/algorithms/dqns/qr_dqn.jl [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    if haskey(batch, :next_legal_actions_mask)
+        l′ = send_to_device(D, batch[:next_legal_actions_mask])
+        avg_q .+= ifelse.(l′, 0.0f0, typemin(Float32))
+    end
+
+    aₜ = argmax(avg_q, dims=2)
+    aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:N-1, 0:0, 0:0)))
+    qₜ = reshape(view(target_q, aₜ), :, batch_size)
+    target =
+    reshape(r, 1, batch_size) .+ γ * reshape(1 .- t, 1, batch_size) .* qₜ
```
Comment (findmyway) [inline src/algorithms/dqns/qr_dqn.jl]: The calculation of target quantiles seems to be incorrect here. Please have a check.

**Review by findmyway — CHANGES_REQUESTED** [review]

File: src/algorithms/dqns/qr_dqn.jl (hunk 8 above)
Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: I have checked the code and the outputs with jupyter notebook and the selection and calculation of values seem to be right to me. 

If there is anything specific, please do let me know.

Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: OK, I was meant to let you figure it out by yourself. Could you share more details about how you check the code and outputs? 

From the algorithm you pasted above:

image

But obviously you are not calculating the correct quantile value here. This makes me suspect that if you have fully understood the algorithm before implementing it.

Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: I am sorry. I wasn't able to understand much of the notation given in the paper so I used https://github.com/ku2482/fqf-iqn-qrdqn.pytorch for reference. 

I just made sure that the code worked as intended by getting the outputs and checking if the outputs are as expected.

Clearly, there is something wrong with the approach. I will try to fix that.

Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: Here, the `j` means the `j`-th quantile, and `a*` is the selected best action.

So to calculate the `target_quantile`, we simply select the the `a*`-th slice along the second dimension of `qₜ`.

Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: In the implemented algorithm:

- I take the average for all the quantiles for the actions (1st dimension) for s'.
- Then take the argmax to find the best action. 
- Then found the target slices indices using at.
- Then used at to select the action slices in Q(s').
- Then calculated target.

This sounds the same as the interpretation of the algorithm.

I am sorry if I got something wrong.

Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: `aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:N-1, 0:0, 0:0)))` I guess this line doesn't work as you expected.

Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: **Works as expected:**
image
image

Comment (findmyway) [reply src/algorithms/dqns/qr_dqn.jl]: That's..., correct though. Sorry that this is the first time I saw this kind of array indexing.

Generally, people would drop the dimension of `avg_q` after `mean`, getting a 2-d matrix, then we find the `aₜ ` which is of size `(batch_size,)`, then we simply read the target_quantile like this: `select_dim(target_q, 2, aₜ)`.

Comment (Mobius1D) [reply src/algorithms/dqns/qr_dqn.jl]: Oh, alright. That makes sense. 

I think what you suggest is more efficient. So, can I make a change?

**Comment by findmyway** [comment]
This should make the implementation easier to read.

https://github.com/Mobius1D/ReinforcementLearningZoo.jl/pull/1

**Comment by findmyway** [comment]
FYI:

image
